Replace catchall kernels with CompoundOp dispatch key.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32782 Replace catchall kernels with CompoundOp dispatch key.**
* #32788 Add XLAPreAutograd key for XLA use cases that need custom autograd.
* #32787 Stop using dispatchTypeId to do checks for tensor list unwrap.
* #32734 Centralize addition of "always on" dispatch keys.
* #32729 Make DispatchKeyGuards accept DispatchKey::Undefined
* #32728 Rename DispatchKey::UndefinedTensorId to Undefined

This eliminates a special case in the dispatcher.  See
https://github.com/pytorch/pytorch/issues/29548 for more
information.

I had to fix the unwrapping of TensorList (discharging a TODO) because after this change it was reporting CompoundOp instead of the correct CPU/CUDA backend.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19628644](https://our.internmc.facebook.com/intern/diff/D19628644)

ezyang committed:
Replace catchall kernels with CompoundOp dispatch key.

This eliminates a special case in the dispatcher.  See
https://github.com/pytorch/pytorch/issues/29548 for more
information.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 252c0a4d:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=252c0a4d50bf75aab3e81582de675fb234365047).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4371323) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92720833) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92720833))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
  git checkout -b <new-branch-name> 
 
HEAD is now at 252c0a4d50 Replace catchall kernels with CompoundOp dispatch key. 
+ git reset --hard 252c0a4d50bf75aab3e81582de675fb234365047 
HEAD is now at 252c0a4d50 Replace catchall kernels with CompoundOp dispatch key. 
+ git merge --allow-unrelated-histories --no-edit --no-ff eab99ab08e03414681ae9e1efc2de01e0ebc49ed 
Auto-merging c10/core/DispatchKey.h 
CONFLICT (content): Merge conflict in c10/core/DispatchKey.h 
Removing aten/src/ATen/detail/ScalarTypeConversions.h 
Auto-merging aten/src/ATen/core/dispatch/DispatchKeyExtractor.h 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

ezyang committed:
Update on "Replace catchall kernels with CompoundOp dispatch key."


This eliminates a special case in the dispatcher.  See
https://github.com/pytorch/pytorch/issues/29548 for more
information.

I had to fix the unwrapping of TensorList (discharging a TODO) because after this change it was reporting CompoundOp instead of the correct CPU/CUDA backend.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace catchall kernels with CompoundOp dispatch key."


This eliminates a special case in the dispatcher.  See
https://github.com/pytorch/pytorch/issues/29548 for more
information.

I had to fix the unwrapping of TensorList (discharging a TODO) because after this change it was reporting CompoundOp instead of the correct CPU/CUDA backend.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Replace catchall kernels with CompoundOp dispatch key."


This eliminates a special case in the dispatcher.  See
https://github.com/pytorch/pytorch/issues/29548 for more
information.

I had to fix the unwrapping of TensorList (discharging a TODO) because after this change it was reporting CompoundOp instead of the correct CPU/CUDA backend.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
